File: app/routers/scraper.py
import json
import logging
import re

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import Professor
from app.schemas import MatchScoreInput, ProfessorCreate, ScraperInput
from app.services.llm import extract_text_content, invoke_llm, parse_json_from_text

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/api/scraper", tags=["scraper"])


@router.post("/fetch", response_model=dict)
async def fetch_professors(data: ScraperInput, db: AsyncSession = Depends(get_db)):
    prompt = f"""Please search and list relevant professor information based on the following:

University: {data.university}
Department: {data.department}
Research Area: {data.research_area}

Return a JSON array containing at least 5 professors. Each professor should include:
- name: Professor name
- university: University name
- department: Department name
- researchAreas: Research areas (comma separated)
- recentPublications: Recent paper titles and abstracts (semicolon separated)
- homepageUrl: Personal homepage URL (if available)
- googleScholarUrl: Google Scholar URL (if available)
- labWebsiteUrl: Lab website URL (if available)

Return format:
[
  {{
    "name": "Professor Name",
    "university": "University Name",
    "department": "Department Name",
    "researchAreas": "Area1, Area2",
    "recentPublications": "Paper1 abstract; Paper2 abstract",
    "homepageUrl": "URL",
    "googleScholarUrl": "URL",
    "labWebsiteUrl": "URL"
  }}
]

Return ONLY the JSON array, no other text."""

    try:
        response = await invoke_llm(
            messages=[{"role": "user", "content": prompt}],
        )

        content = extract_text_content(response)
        if not content:
            raise ValueError("Invalid response from LLM")

        json_match = re.search(r"\[[\s\S]*\]", content)
        if not json_match:
            raise ValueError("Failed to parse professor data from LLM response")

        professors_data = json.loads(json_match.group(0))

        created_professors = []
        for prof in professors_data:
            professor = Professor(
                name=prof.get("name", ""),
                university=prof.get("university", ""),
                department=prof.get("department", ""),
                research_areas=prof.get("researchAreas", ""),
                recent_publications=prof.get("recentPublications", ""),
                homepage_url=prof.get("homepageUrl", ""),
                google_scholar_url=prof.get("googleScholarUrl", ""),
                lab_website_url=prof.get("labWebsiteUrl", ""),
            )
            db.add(professor)
            await db.flush()
            created_professors.append({"id": professor.id, **prof})

        await db.commit()

        return {
            "success": True,
            "count": len(created_professors),
            "professors": created_professors,
        }
    except Exception as e:
        logger.error(f"Error fetching professors: {e}")
        raise


@router.post("/match-score", response_model=dict)
async def generate_match_score(data: MatchScoreInput):
    prompt = f"""Please calculate the match score (0-100) between the user and professor based on the following:

User Background:
{data.user_background}

Professor Research:
{data.professor_research}

Evaluate the following aspects:
1. Research direction alignment
2. Academic background relevance
3. Research interest consistency

Return a JSON object containing:
- score: Overall match score (0-100)
- reasoning: Brief explanation
- strengths: Matching strengths (array)
- gaps: Areas for improvement (array)

Return format:
{{
  "score": 85,
  "reasoning": "User's research background highly aligns with professor's research direction...",
  "strengths": ["strength1", "strength2"],
  "gaps": ["gap1", "gap2"]
}}

Return ONLY the JSON object, no other text."""

    try:
        response = await invoke_llm(
            messages=[{"role": "user", "content": prompt}],
        )

        content = extract_text_content(response)
        if not content:
            raise ValueError("Invalid response from LLM")

        match_data = parse_json_from_text(content)
        if not match_data:
            raise ValueError("Failed to parse match score from LLM response")

        return match_data
    except Exception as e:
        logger.error(f"Error generating match score: {e}")
        raise

File: server/routers/scraper.ts
import { z } from "zod";
import { protectedProcedure, router } from "../_core/trpc";
import { invokeLLM } from "../_core/llm";
import { createProfessor } from "../db";

export const scraperRouter = router({
  fetchProfessors: protectedProcedure
    .input(
      z.object({
        university: z.string(),
        department: z.string(),
        researchArea: z.string(),
      })
    )
    .mutation(async ({ input }) => {
      const prompt = `请根据以下信息搜索并列出相关的教授信息：

大学：${input.university}
部门：${input.department}
研究方向：${input.researchArea}

请返回一个 JSON 数组，包含至少 5 位教授的信息。每位教授应包含以下字段：
- name: 教授姓名
- university: 所在大学
- department: 所在部门
- researchAreas: 研究方向（用逗号分隔）
- recentPublications: 近期论文标题和摘要（用分号分隔）
- homepageUrl: 个人主页 URL（如果有）
- googleScholarUrl: Google Scholar URL（如果有）
- labWebsiteUrl: 实验室网站 URL（如果有）

返回格式：
[
  {
    "name": "教授姓名",
    "university": "大学名称",
    "department": "部门名称",
    "researchAreas": "研究方向1, 研究方向2",
    "recentPublications": "论文1摘要; 论文2摘要",
    "homepageUrl": "URL",
    "googleScholarUrl": "URL",
    "labWebsiteUrl": "URL"
  }
]

请只返回 JSON 数组，不要包含其他文本。`;

      try {
        const response = await invokeLLM({
          messages: [
            {
              role: "user",
              content: prompt,
            },
          ],
        });

        // Extract text content from response
        const textContent = response.choices[0]?.message.content;
        if (!textContent || typeof textContent !== "string") {
          throw new Error("Invalid response format from LLM");
        }

        // Parse the response as JSON
        const jsonMatch = textContent.match(/\[[\s\S]*\]/);
        if (!jsonMatch) {
          throw new Error("Failed to parse professor data from LLM response");
        }

        const professorsData = JSON.parse(jsonMatch[0]);

        // Save professors to database
        const createdProfessors = [];
        for (const prof of professorsData) {
          const id = await createProfessor({
            name: prof.name,
            university: prof.university,
            department: prof.department,
            researchAreas: prof.researchAreas,
            recentPublications: prof.recentPublications,
            homepageUrl: prof.homepageUrl,
            googleScholarUrl: prof.googleScholarUrl,
            labWebsiteUrl: prof.labWebsiteUrl,
          });
          createdProfessors.push({ id, ...prof });
        }

        return {
          success: true,
          count: createdProfessors.length,
          professors: createdProfessors,
        };
      } catch (error) {
        console.error("Error fetching professors:", error);
        throw error;
      }
    }),

  generateMatchScore: protectedProcedure
    .input(
      z.object({
        userBackground: z.string(),
        professorResearch: z.string(),
      })
    )
    .mutation(async ({ input }) => {
      const prompt = `请根据以下信息计算用户与教授的匹配度评分（0-100）：

用户背景：
${input.userBackground}

教授研究方向：
${input.professorResearch}

请评估以下几个方面的匹配度：
1. 研究方向的契合度
2. 学术背景的相关性
3. 研究兴趣的一致性

请返回一个 JSON 对象，包含：
- score: 总体匹配度评分（0-100）
- reasoning: 评分原因（简要说明）
- strengths: 匹配的优势（数组）
- gaps: 需要改进的地方（数组）

返回格式：
{
  "score": 85,
  "reasoning": "用户的研究背景与教授的研究方向高度契合...",
  "strengths": ["优势1", "优势2"],
  "gaps": ["需改进1", "需改进2"]
}

请只返回 JSON 对象，不要包含其他文本。`;

      try {
        const response = await invokeLLM({
          messages: [
            {
              role: "user",
              content: prompt,
            },
          ],
        });

        const textContent = response.choices[0]?.message.content;
        if (!textContent || typeof textContent !== "string") {
          throw new Error("Invalid response format from LLM");
        }

        const jsonMatch = textContent.match(/\{[\s\S]*\}/);
        if (!jsonMatch) {
          throw new Error("Failed to parse match score from LLM response");
        }

        const matchData = JSON.parse(jsonMatch[0]);
        return matchData;
      } catch (error) {
        console.error("Error generating match score:", error);
        throw error;
      }
    }),
});
